### PR TITLE
bacpop-100 External clusters

### PR DESCRIPTION
**Pull request by EmmaLRussell**
This branch ensures that clusters defined in the GPSC external clusters file are returned for all analysis results, rather than internal (poppunk) clusters. 

This branch is deployed to http://beebop-dev.dide.ic.ac.uk
Nick has had a look at this and declared it to be working as expected. 

Key changes:
- added the external clusters csv file to the resources folder
- pass that filepath to poppunk assign and `generate_visualisation` calls
- use the external_clusters csv file (rather than clusters.csv) generated by the assign call to build the API response
- also use the external_clusters file as `previous_query_clustering` parameter of the poppunk `generate_visualisation` calls
- when constructing assign responses, we want to assign a single external cluster to each sample, whereas the external clusters files have multiple cluster numbers, separated by semicolons. As requested by Nick, we assign lowest cluster number for each sample. 
- we also put "GPSC" prefix on the clusters, to be displayed in the front end
- there is one place where we still need to deal with internal poppunk clusters, and that's when passing the include file to the `generate_visualisation` call for microreact. That include file is generated by the `summarise_clusters` call, and is tagged with the internal cluster. So we save a mapping of external cluster labels to internal cluster number when we do the assignment, for later use to fetch this file on create microreact. 
- there was as issue when mapping clusters to network components (in `utils.generate_mapping`) - network components get arbitrary ids which we need to map to clusters ourselves. Previously we were abe to assume that a cluster which shared any sample with a component could be mapped to that component. However, the network now seems to be generating some non-cluster single-sample components which we definitely don't want to map to. However, we can't require that components and clusters be exact matches either (possibly because of the lowest number assignment?) so instead I've put in some very inefficient best-match code which seems to work well enough. Nick thinks this is an issue with the poppunk code, and is going to look into it, so we should be able to revert this later. 
- I've tried to rationalise the tests a bit so that they share code to do assignment and run network, and so tests don't rely on each others' output.

No changes were required in beebop. The GPSC prefix makes the network tab labels wrap, which looks a bit weird, but I think we can sort that out in the new UI!

To test:
You should find that the attached files (provided by Nick): 
[19183_3#105.fasta.txt](https://github.com/bacpop/beebop_py/files/14613272/19183_3.105.fasta.txt) gets assigned to GPSC22
[22841_2#137.fasta.txt](https://github.com/bacpop/beebop_py/files/14613274/22841_2.137.fasta.txt) gets assigned to GPSC9

..and that network and download output look right, and are tagged with the external cluster number. 
(NB remove .txt suffix to upload to beebop - github won't attach a .fasta file!)

### CODE REVIEW

File: beebop/utils.py [hunk 1]
```diff
@@ -172,3 +218,58 @@ def add_query_ref_status(fs: PoppunkFileStore,
...
+    there are multiple external clusters listed for a sample, the lowest
+    cluster number is returned]
+
+    :param external_clusters_file: [filename of the project's external clusters
+        file]
+    :param hashes_list: [list of sample hashes to find samples for]
+    :return dict: [dict of sample hash to lowest numbered-cluster for that
+        sample]
+    """
+    remaining_hashes = hashes_list[:]
```
Comment (absternator) [inline beebop/utils.py:252]: if this is just making a copy it should be named accordingly `hashes_list_copy`

File: docker/Dockerfile [hunk 2]
```diff
@@ -4,7 +4,7 @@ RUN conda config --append channels conda-forge
 RUN conda config --append channels bioconda
 
 RUN conda install python=3.10
-RUN conda install poppunk
+RUN conda install poppunk=2.6.5
```
Comment (absternator) [inline docker/Dockerfile:7]: what was the reason to lock to this version? just wondering if there is updates to popunk we will need to update this every time

**Review by absternator — COMMENTED** [review]
Awsome work all seems to be working great!! just some tiny commnets mainly naming and for clarity

File: beebop/app.py [hunk 3]
```diff
@@ -82,14 +82,15 @@ def generate_zip(fs: PoppunkFileStore,
     :return BytesIO: [memory file]
     """
     memory_file = BytesIO()
+    cluster_no = cluster_no_from_label(cluster)
```
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply beebop/app.py:85]: I think I inherited 'no' meaning 'number' from somewhere else in the beebop_py code, but I'll change to 'num', you're right it is clearer..

File: beebop/assignClusters.py [hunk 4]
```diff
@@ -70,13 +71,28 @@ def get_clusters(hashes_list: list,
...
+        get_external_clusters_from_file(external_clusters_file, hashes_list)
     result = {}
-    for i, (name, cluster) in enumerate(zip(queries_names, queries_clusters)):
+    for i, (name, cluster) in enumerate(external_clusters.items()):
         result[i] = {
             "hash": name,
             "cluster": cluster
         }
 
+    # save a mapping of PopPUNK clusters to external clusters which
```
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply beebop/assignClusters.py:85]: Will extract this save now, maybe further split can be delayed til after the demo.

[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply beebop/assignClusters.py:85]: Just noticed the comment is misleading as the mapping goes the other way, will fix that too.

File: beebop/filestore.py [hunk 5]
```diff
@@ -88,6 +90,14 @@ def output_cluster(self, p_hash) -> str:
         """
         return str(PurePath(self.output(p_hash), "cluster_results.pickle"))
 
+    def external_to_poppunk_clusters(self, p_hash) -> str:
```
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply beebop/filestore.py:93]: It's a mapping of external clusters labels to internal poppunk clusters - we only need this for the include file, see note above:
 > there is one place where we still need to deal with internal poppunk clusters, and that's when passing the include file to the generate_visualisation call for microreact. That include file is generated by the summarise_clusters call, and is tagged with the internal cluster. So we save a mapping of external cluster labels to internal cluster number when we do the assignment, for later use to fetch this file on create microreact.

File: docker/Dockerfile (hunk 2 above)
Comment (EmmaLRussell) [reply docker/Dockerfile:7]: it didn't seem to update to latest version for me, but I think maybe that was before the new version had the main label. 
Will try it again without.

Comment (EmmaLRussell) [reply docker/Dockerfile]: Actually, I think the issue was that the new version had only gone on bioconda, not conda forge. So I'm now specifying that channel, as I think that will always have the latest.

**Review by absternator — APPROVED** [review]
LGTM

**Pull request merged**